jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java

```diff
@@ -1051,9 +1106,13 @@ else if (_counts.getHi() == Integer.MIN_VALUE)
                             long idleTimeout = getIdleTimeout();
                             if (idleTimeout > 0 && getThreads() > _minThreads)
                             {
-                                long last = _lastShrink.get();
+                                long last;
                                 long now = NanoTime.now();
-                                if (NanoTime.millisElapsed(last, now) > idleTimeout && _lastShrink.compareAndSet(last, now))
+                                long itNanos = TimeUnit.MILLISECONDS.toNanos(idleTimeout);
```

Comment:
https://github.com/magibney/jetty.project/commit/dfab46bb3fff412683fbb847e08c9da649fa4088#r100644358
>why isn't this just:
```java
    long last = _lastShrink.get();
    long now = NanoTime.now();
    if (NanoTime.nanosElapsed(last, now) > getShrinkInterval() && _lastShrink.compareAndSet(last, now))
```

Reply:
I'm not sure I understand the concern about doing the `compareAndSet` in a loop.   If it fails the CaS, then the value has been updated to the `now` as seen by another thread.  That might be a little before or after the now this thread just calculated, but it should not be so different as to be another `getShrinkInterval()` away.   If you fail the CaS and loop, then the new expiry will be checked and the elapsed time will be seen to be less that the shrink interval, so the loop will exit.       If the machine is so loaded that it takes longer... then having an idle thread spin on this operation is the last thing we should be using CPU on :)

Reply:
I would go for simpler logic here and if we failed the CaS, then pass the shrink timeout to `idleJobPoll` below.

Reply:
08c65bc10db32218093737dde3b4ed87bd45cc58 incorporates your suggestions but in a slightly different way. I'm fine with however you want to proceed, but my inclination would be to either keep the CaS loop (in its updated/clarified form) or go completely the other direction and not worry about it (i.e., continue to `idleJobPoll()` at rate of once per `idleTimeout` and just live with the unlikely possibility of delayed pool shrinkage).

If we were to pass `shrinkInterval` to `idleJobPool()`, we'd either have to do that all the time (potentially approaching busy-wait for smaller `shrinkIntervals`), or we'd have to only pass `shrinkInterval` in the event that the CaS has just failed (certainly doable, but undermines the goal of code simplicity).

The main reason I added CaS loop was to make behavior more consistent for testing (practically speaking, I'm confident the pool will shrink well enough with or without CaS loop); and passing `shrinkInterval` to `idleJobPoll()`, although it would allow faster pool shrinkage in cases of extreme contention, doesn't really help wrt consistency.

Reply:
I tested this a little, and you were right to begin with @gregw -- failed CaS is just a non-issue. I added a note in case it raises any eyebrows in the future, but if you think that's overkill the note could be removed. I think sticking with `idleTimeout` for `idleJobPoll()` will also be perfectly fine.

Made these changes in 81bcdec2565de1bc0eb250e246b67aa16bfe52db; I left the lazy assignment of `last` and `siNanos` local variables; lmk if you'd prefer to unconditionally initialize those, that's fine.